### PR TITLE
[WIP]Utxo fragmentation

### LINKED ISSUE #184: Concurrent swaps: Fragment UTXOS
Each trade locks an utxo, therefore beside the size of the trade, the number of requests we can serve is linear with the number of nominal utxo we have.

In order to accommodate multiple concurrent trades at the same time, an easy solution is to fragment the utxos at deposit time, in many smaller utxos, so its easier to find utxo available for the trade.

This could be done either as external script or embedded into the daemon. 

This would means creating an ephemeral address for the user to deposit his UTXO, we would detect this, create a transaction that spends it and creates a number of smaller outputs versus the address of the given DepostiAddress rpc result.

The ratios of the amount of the outputs should reflect a scaled distribution, to be defined.

### PR DESCRIPTION
**Pull request by sekulicd**
This is new PR which should replace #194.

It merges master on top of previous PR and adds claim unspents automatically.
It is in WIP since it needs to be better tested alongside with Wallet restore and other RPCs that could be affected with high number of generated addresses and unspents (eg. RestoreUtxos)

This closes #184 

@tiero @altafan please review.

### CODE REVIEW

File: cmd/tdex/depositmarket_test.go [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+		pair AssetValuePair
+	}
+	tests := []struct {
+		name  string
+		args  args
+		want  []uint64
+		want1 []uint64
+	}{
+		{
+			name: "test1",
```
Comment (altafan) [inline cmd/tdex/depositmarket_test.go:20]: names should be there to let one understand what a test does. Can you find a better one?

Also I think that we should increase the test vectors to cover also edge cases.
Let's do this now instead of (hopefully not) experiencing issues with running daemons and being forced later to debug -> fix -> create such kind of tests.

File: cmd/tdex/depositfee.go [hunk 2]
```diff
@@ -27,17 +54,256 @@ func depositFeeAction(ctx *cli.Context) error {
...
+		resp, err := explorerSvc.BroadcastTransaction(txHex)
+		if err != nil {
+			log.Warn(err)
+			log.Info("transaction broadcast retry")
+			continue retry
+		}
+		log.Info(resp)
+		txID = resp
+		break retry
+	}
```
Comment (altafan) [inline cmd/tdex/depositfee.go:150]: why retrying? It's unlikely that retrying to broadcast a tx makes it accepted in mempool if the first time it didn't. If the tx gets rejected, it's because it's invalid, ie. break the Elements protocol, and retrying won't make any difference. Am I wrong?

File: cmd/tdex/depositfee.go [hunk 3]
```diff
@@ -27,17 +54,256 @@ func depositFeeAction(ctx *cli.Context) error {
...
+	outpoints := make([]*pboperator.TxOutpoint, 0, len(outputs))
+	for i := 0; i < len(outputs); i++ {
+		outpoints = append(outpoints, &pboperator.TxOutpoint{
+			Hash:  txID,
+			Index: int32(i),
+		})
+	}
+
+	//wait one min so that tx get confirmed
+	time.Sleep(5 * time.Second)
```
Comment (altafan) [inline cmd/tdex/depositfee.go:161]: It must be at least 1 minute of waiting time in mainnet!

File: cmd/tdex/depositmarket.go [hunk 4]
```diff
@@ -2,26 +2,55 @@ package main
...
+			Value: false,
+		},
 		&cli.StringFlag{
-			Name:  "quote_asset",
-			Usage: "the base asset hash of an existent market",
-			Value: "",
+			Name:  "explorer",
+			Usage: "explorer endpoint url",
+			Value: "http://127.0.0.1:3001",
 		},
```
Comment (altafan) [inline cmd/tdex/depositmarket.go:54]: why is this needed? can't we get the explorer from config ad done in depositfee?

File: cmd/tdex/depositmarket.go [hunk 5]
```diff
@@ -31,29 +60,553 @@ var depositmarket = cli.Command{
...
+	baseAssetOpt := ""
+	if quoteAssetOpt != "" {
+		baseAssetOpt = baseAssetKey
+	}
+
+	if fragmentationDisabled {
+		numOfAddresses := int64(1)
+		if ctx.Int64("num_of_addresses") > 1 {
+			numOfAddresses = ctx.Int64("num_of_addresses")
+		}
```
Comment (altafan) [inline cmd/tdex/depositmarket.go:96]:
```suggestion
		numOfAddresses := ctx.Int64("num_of_addresses")
```

File: cmd/tdex/depositmarket.go [hunk 6]
```diff
@@ -31,29 +60,553 @@ var depositmarket = cli.Command{
...
+		break
+	}
+
+	outpoints := make([]*pboperator.TxOutpoint, 0, len(outputs))
+	for i := 0; i < len(outputs); i++ {
+		outpoints = append(outpoints, &pboperator.TxOutpoint{
+			Hash:  txID,
+			Index: int32(i),
+		})
+	}
```
Comment (altafan) [inline cmd/tdex/depositmarket.go:203]:
```suggestion
	outpoints := make([]*pboperator.TxOutpoint, outsLen, outsLen)
	for i:= 0; i < outsLen; i++ {
		outpoints[i] = &pboperator.TxOutpoint{
			Hash:  txID,
			Index: int32(i),
		})
	}
```

File: cmd/tdex/depositfee.go [hunk 7]
```diff
@@ -27,17 +54,256 @@ func depositFeeAction(ctx *cli.Context) error {
...
+		break retry
+	}
+
+	outpoints := make([]*pboperator.TxOutpoint, 0, len(outputs))
+	for i := 0; i < len(outputs); i++ {
+		outpoints = append(outpoints, &pboperator.TxOutpoint{
+			Hash:  txID,
+			Index: int32(i),
+		})
+	}
```
Comment (altafan) [inline cmd/tdex/depositfee.go:158]:
```suggestion
	outpoints := make([]*pboperator.TxOutpoint, len(outputs), len(outputs))
	for i := 0; i < len(outputs); i++ {
		outpoints[i] = &pboperator.TxOutpoint{
			Hash:  txID,
			Index: int32(i),
		}
	}
```

File: cmd/tdex/depositmarket.go [hunk 8]
```diff
@@ -31,29 +60,553 @@ var depositmarket = cli.Command{
...
+
+	//wait one min so that tx get confirmed
+	time.Sleep(5 * time.Second)
```
Comment (altafan) [inline cmd/tdex/depositmarket.go:206]: It must be at least 1 minute of waiting time in mainnet!

File: cmd/tdex/depositmarket.go [hunk 9]
```diff
@@ -31,29 +60,553 @@ var depositmarket = cli.Command{
...
+	})
+
+	sort.Slice(quoteAssetFragments, func(i, j int) bool {
+		return quoteAssetFragments[i] < quoteAssetFragments[j]
+	})
+
+	//if there is rest, created when calculating percentage,
+	//add it to last fragment
+	if baseSum != pair.BaseValue {
+		baseRest := pair.BaseValue - baseSum
```
Comment (altafan) [inline cmd/tdex/depositmarket.go:312]: could it happen that baseSum is actually > pair.BaseValue? it shouldn't ever happen I guess, but I don't see any code that enforces this.

**Review by altafan — CHANGES_REQUESTED** [review]

File: cmd/tdex/depositfee.go [hunk 10]
```diff
@@ -2,22 +2,49 @@ package main
...
+		},
+		&cli.StringFlag{
+			Name:  "explorer",
+			Usage: "explorer endpoint url",
+			Value: "http://127.0.0.1:3001",
+		},
+		&cli.IntFlag{
+			Name:  "num_of_addresses",
+			Usage: "the number of addresses to retrieve",
+		},
```
[the remark this answers is not in the document]
Comment (sekulicd) [reply cmd/tdex/depositfee.go:42]: can you refer to the duplicate?

File: cmd/tdex/depositfee.go [hunk 11]
```diff
@@ -27,17 +54,256 @@ func depositFeeAction(ctx *cli.Context) error {
...
 
-	numOfAddresses := ctx.Int64("num_of_addresses")
+	net, baseAssetKey := getNetworkAndBaseAssetKey(ctx.String("network"))
+	fragmentationDisabled := ctx.Bool("no-fragment")
+
+	if fragmentationDisabled {
+		numOfAddresses := int64(1)
+		if ctx.Int64("num_of_addresses") > 1 {
+			numOfAddresses = ctx.Int64("num_of_addresses")
+		}
```
[the remark this answers is not in the document]
Comment (sekulicd) [reply cmd/tdex/depositfee.go:64]: Done

File: cmd/tdex/depositfee.go [hunk 12]
```diff
@@ -27,17 +54,256 @@ func depositFeeAction(ctx *cli.Context) error {
...
+
+			for _, v := range unspents {
+				valuePerAsset[v.Asset()] += v.Value()
+			}
+
+			switch len(valuePerAsset) {
+			case 1:
+				for k, v := range valuePerAsset {
+					if k == baseAssetKey {
+						if v < MinFee {
```
[the remark this answers is not in the document]
Comment (sekulicd) [reply cmd/tdex/depositfee.go:262]: Done

File: cmd/tdex/depositfee.go [hunk 13]
```diff
@@ -27,17 +54,256 @@ func depositFeeAction(ctx *cli.Context) error {
...
+	BaseAsset string
+}
+
+// findBaseAssetsUnspents polls blockchain until base asset unspent is noticed
+func findBaseAssetsUnspents(
+	randomWallet *trade.Wallet,
+	explorerSvc explorer.Service,
+	baseAssetKey string,
+) (
+	BaseAssetPair,
```
[the remark this answers is not in the document]
Comment (sekulicd) [reply cmd/tdex/depositfee.go:234]: Done

File: cmd/tdex/depositfee.go [hunk 14]
```diff
@@ -27,17 +54,256 @@ func depositFeeAction(ctx *cli.Context) error {
...
+		baseAssetKey,
+	)
+
+	log.Info("calculating fragments ...")
+	baseFragments := fragmentFeeUnspents(
+		baseAssetPair.BaseValue,
+		MinFee,
+		MaxNumOfOutputs,
+	)
+	log.Infof("base fragments: %v", baseFragments)
```
[the remark this answers is not in the document]
Comment (sekulicd) [reply cmd/tdex/depositfee.go:101]: Done

File: cmd/tdex/depositmarket.go (hunk 5 above)
Comment (sekulicd) [reply cmd/tdex/depositmarket.go:96]: Done

File: cmd/tdex/depositmarket.go (hunk 8 above)
Comment (sekulicd) [reply cmd/tdex/depositmarket.go:206]: yeap, i left comment, this was just for faster testing :)
Done

Comment (sekulicd) [reply cmd/tdex/depositmarket.go:206]: @tiero @altafan should i put here 65sec maybe?

File: cmd/tdex/depositmarket.go (hunk 6 above)
Comment (sekulicd) [reply cmd/tdex/depositmarket.go:203]: We had this [discussion before](https://github.com/TDex-network/tdex-daemon/pull/319#discussion_r601619889), I dont think this is significant improvement if at all.
@tiero what do you think about this? should we use build in append or go with above way  of populating slice for known length?

File: cmd/tdex/depositmarket.go (hunk 9 above)
Comment (sekulicd) [reply cmd/tdex/depositmarket.go:312]: It cant, cause baseSum is sum of parts(percentages) of pair.BaseValue

**Comment by sekulicd** [comment]
@altafan InitWallet gets stuck(i stopped daemon after few minutes) after i run depositfee and depositmarket commands which in the end claims unspents for market and fee account.
I think it has something to do with Circuit Braker.

Steps to reproduce:
1. clean data dir
2. start daemon, init walet, unlock wallet
3. run depositfee(faucet) and depositmarket(faucet and mint)
4. call Open Market RPC with quote assed from depositmarket output
5. restart daemon

Can u try to reproduce?

File: cmd/tdex/depositmarket.go (hunk 8 above)
Comment (tiero) [reply cmd/tdex/depositmarket.go]: yes better to account for 5 seconds of tolerance

**Comment by altafan** [comment]
Closing in favour of #332 .
